### PR TITLE
feat:add search/export/query function

### PR DESCRIPTION
**Pull request by zbjdonald**
1. Add search api wrapper.
  2. Get/Export saved question data.

### CODE REVIEW

**Comment by vvaezian** [comment]
@zbjdonald  Thanks for the PR.
I'll review it on the weekend.

File: metabase_api/metabase_api.py [hunk 1]
```diff
@@ -778,8 +778,64 @@ def copy_collection(self, source_collection_name=None, source_collection_id=None
...
+    card_id -- card_id, int type
+    collection_id -- collection_id
+    collection_name -- collection_name
+    """
+    if export_format not in [None, 'csv', 'json', 'xlsx']:
+      raise ValueError('{} format not supported. Supported format are csv, json, xlsx'.format(export_format))
+    if [card_id, card_name].count(None) == 2:
+      raise ValueError('Either card_id or card_name should be available')
+    if card_id is None:
+      if collection_id is None and collection_name is None:
```
Comment (vvaezian) [inline metabase_api/metabase_api.py:820]: - Why not using `get_item_id` for this case as well? Is there a need for the `search` function that you defined?

File: metabase_api/metabase_api.py [hunk 2]
```diff
@@ -778,8 +778,64 @@ def copy_collection(self, source_collection_name=None, source_collection_id=None
...
+        search_ret = self.search(card_name, item_type='card', single_ret=True)
+        card_id = search_ret['id']
```
Comment (vvaezian) [inline metabase_api/metabase_api.py:822]: Why not using `get_item_id` for this case as well? Is there a need for the `search` function that you defined?

Comment (zbjdonald) [reply metabase_api/metabase_api.py:822]: Pull all collections and filter locally may slow than search API if there are lots of collections. I also find there is no search API wrapper in your code, So I add the "search" function.

Comment (vvaezian) [reply metabase_api/metabase_api.py:822]: That's a good point regarding performance. But the fix should be done on the underlying functions (e.x. `get_item_id`) so other functions can also benefit from the improved performance. 
I'll merge your commit and then make some changes to reflect what has been discussed, and to apply some generalizations and standardization.

Comment (zbjdonald) [reply metabase_api/metabase_api.py:822]: Ok. When you finished, I will add a “visual_setting” option for the query method.

File: metabase_api/metabase_api.py [hunk 3]
```diff
@@ -778,8 +778,64 @@ def copy_collection(self, source_collection_name=None, source_collection_id=None
...
                         destination_collection_id=destination_parent_collection_id,
                         destination_pulse_name=destination_pulse_name)
-  
-  
+
+
+  def search(self, q, item_type=None, archived=False, single_ret=False):
+    """
+    Search content and filter item type. Note: endpoint /api/search fall behind with web app version.
+
```
Comment (vvaezian) [inline metabase_api/metabase_api.py:786]: One last question: What do you mean by "endpoint /api/search fall behind with web app version"?

Comment (zbjdonald) [reply metabase_api/metabase_api.py:786]: Metabase web API provides filter content type param, such as 'card' type, whereas endpoint /api/search does not support this filter param.
**Web API:** 
image
**API in their docs:**
image

Comment (vvaezian) [reply metabase_api/metabase_api.py:786]: Where it says `&type=card` if you look at the `response` tab, you see that the response contains all types. So `type` is apparently something used for the frontend. Under the hood all the results have actually been returned.

**Comment by vvaezian** [comment]
I made some changes to make them inline with other functions. The functions are called `search` and `get_card_data`.

It took longer that expected because I was considering how to incorporate the changes.  
The idea about rewriting all the auxiliary functions using the `search` function is something I'll consider in the future.

**Pull request merged**